proc-macros/src/render_client.rs

```diff
@@ -68,6 +68,41 @@ impl RpcDescription {
 		Ok(trait_impl)
 	}
 
+	fn patch_result_error(&self, ty: &syn::Type) -> syn::Type {
```

Comment:
I had a play and wonder whether this would be a clearer approach to take:

```rust
fn return_result_type(&self, ty: &syn::Type) -> TokenStream2 {
	// We expect a valid type path.
	let syn::Type::Path(type_path) = ty else  {
		return quote_spanned!(ty.span() => compile_error!("Expecting something like 'Result<Foo, Err>' here. (1)"));
	};

	// The path (eg std::result::Result) should have a final segment like 'Result'.
	let Some(type_name) = type_path.path.segments.last() else {
		return quote_spanned!(ty.span() => compile_error!("Expecting this path to end in something like 'Result<Foo, Err>'"));
	};

	// Get the generic args eg the <T, E> in Result<T, E>.
	let PathArguments::AngleBracketed(AngleBracketedGenericArguments { args, .. }) = &type_name.arguments else {
		return quote_spanned!(ty.span() => compile_error!("Expecting something like 'Result<Foo, Err>' here, but got no generic args (eg no '<Foo,Err>')."));
	};

	if type_name.ident == "Result"{
		// Result<T, E> should have 2 generic args.
		if args.len() != 2 {
			return quote_spanned!(args.span() => compile_error!("Expecting two generic args here."));
		}
	} else if type_name.ident == "RpcResult" {
		// RpcResult<T> (an alias we export) should have 1 generic arg.
		if args.len() != 1 {
			return quote_spanned!(args.span() => compile_error!("Expecting two generic args here."));
		}
	} else {
		// Any other type name isn't allowed.
		return quote_spanned!(type_name.span() => compile_error!("The response type should be Result or RpcResult"));
	}

	// The first generic arg is our custom return type.
	// Return a new Result with that custom type and jsonrpsee::core::Error:
	let custom_return_type = args.first().unwrap();
	let error_type = self.jrps_client_item(quote! { core::Error });
	quote!(std::result::Result<#custom_return_type, #error_type>)
}
```

Aside from being documented, it checks for both Result and RpcResult (and verified the number of generics), and returns nicer compile errors which highlight the offending return type when it's wrong.

Instead of modifying the type I just looked for the generic param and then return a new Result type which uses it and the standard error.

Needs testing properly though but it compiles ok!

Reply:
Yeah, it is a different approach; when choosing between the two I figured I'd rather impose as few restrictions as possible, and thus tried to surgically target the `Result`'s error. My rationale for going this way is, for instance, I'm not certain at all that types other than `RpcResult` and `Result<_, _>` are not allowed... But after a quick look through the tests, I can tell it is the case.
So, yeah, there is definitely no such restriction explicitly anywhere just yet, and I wasn't sure it would be the right way of introducing it.

There's also a concern with supporting multiple paths to [`RpcResult`] - but I've implemented a nice helper for handling paths for [`Result`] that would work with [`RpcResult`] too.

Reply:
Tried it! One unexpected artifact is this UI warning:

```
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
warning: unused import: `core::RpcResult`
 --> .../jsonrpsee/proc-macros/tests/ui/correct/only_client.rs:3:17
  |
3 | use jsonrpsee::{core::RpcResult, proc_macros::rpc};
  |                 ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
```

This comes from this code:

```rust
//! Example of using proc macro to generate working client and server.

use jsonrpsee::{core::RpcResult, proc_macros::rpc};

#[rpc(client)]
pub trait Rpc {
	#[method(name = "foo")]
	async fn async_method(&self, param_a: u8, param_b: String) -> RpcResult<u16>;

	#[method(name = "bar")]
	fn sync_method(&self) -> RpcResult<u16>;

	#[subscription(name = "subscribe", item = String)]
	fn sub(&self);
}

fn main() {}
```

If we rewrite the code to use `Result` explicitly it might actually make `RpcResult` unused! Good thing there are UI tests in this project, here...

Reply:
What if user has a custom `Result` type for whatever reason? Shall we replace it with `std::result::Result`? It is a pretty edge case, yet I'd suggest that we pass through that type too, and just alter the last argument. That would be a bit less surprising for someone injecting custom `Result`s.

Reply:
I think I have a version that covers all concerns now, please take a look. It does error rewrite again to be more permissive but provides those nice compile errors (for which I need to add tests actually!).

Reply:
> What if user has a custom `Result` type for whatever reason? Shall we replace it with `std::result::Result`? It is a pretty edge case, yet I'd suggest that we pass through that type too, and just alter the last argument. That would be a bit less surprising for someone injecting custom `Result`s.

I think that is the user has some custom type that isn't something we expect (ie a Result or an RpcResult) then offhand the safest option is to complain, since in the general case we can't possibly know what the structure of said custom type is. @niklasad1 do you have any opinions on this?

Reply:
FYI I had a look at your latest version; it doesn't look any more permissive to me since it still requires the name to be Result or RpcResult (and the Result type still requires 2 args, so it's highly unlikely somebody will be importing a "Result" type alias of that kind anyway). So I still prefer my approach (and avoiding mutating things for clarity) but I'm not super bothered.

Reply:
> I think that is the user has some custom type that isn't something we expect (ie a Result or an RpcResult) then offhand the safest option is to complain, since in the general case we can't possibly know what the structure of said custom type is. @niklasad1 do you have any opinions on this?

Yeah, I agree I think it's better to throw a compile-time error if the error type isn't Result/RpcResult.

Reply:
The case where this solution is more permissive is when the user provides a type called `Result` that has the same shape as the std `Result`, but is a different type. For instance, a struct instead of enum (a case that is unlikely to happen), or a wrapper (much more common, although maybe not for `Result`), for instance for implementing some traits on it that conflict otherwise.